builder/dispatchers.go

```diff
@@ -556,3 +613,53 @@ func volume(b *builder, args []string, attributes map[string]bool, original stri
 	}
 	return nil
 }
+
+// ARG --default=bar --description='my variable' foo
+//
+// Adds the variable foo to the trusted list of variables that can be passed
+// to builder using the --build-arg flag for expansion/subsitution or passing to 'run'.
+// Dockerfile author may optionally set a default value and description of this variable.
+func arg(b *builder, args []string, attributes map[string]bool, original string) error {
+	const (
+		def  = "default"
+		desc = "description"
+	)
+
+	if len(args) != 1 {
+		return fmt.Errorf("ARG requires exactly one argument")
+	}
+
+	// 'arg' accepts the default and description flags
+	flDefault := b.BuilderFlags.AddString(def, "")
+	flDescription := b.BuilderFlags.AddString(desc, "")
+	if err := b.BuilderFlags.Parse(); err != nil {
+		return err
+	}
+
+	arg := args[0]
+
+	// add the arg to allowed list of build-time args from this step on.
+	b.allowedBuildArgs[arg] = make(map[string]string)
+
+	// If there is a default value associated with this arg then add it to the
+	// b.buildArgs if one is not already passed to the builder. The args passed
+	// to builder override the defaut value of 'arg'.
+	if _, ok := b.buildArgs[arg]; !ok {
+		if flDefault.IsUsed() {
+			b.buildArgs[arg] = flDefault.Value
+		}
+	}
+
+	// log the arg and it's default value and description, if any
+	argStr := arg
+	if flDefault.IsUsed() {
+		argStr = fmt.Sprintf("--%s=%s %s", def, flDefault.Value, argStr)
```

Comment:
you may want to use %q here and below instead of %s for the value part, that way its quoted nicely if someone tries to parse it.  Won't get confused with the next word.